Import libraries.

In [54]:
import pandas as pd    # data formatting
import numpy as np     # numeric library
from sklearn.neighbors import KNeighborsClassifier  # machine learning
from sklearn.externals import joblib # exporting model
from sklearn.metrics import confusion_matrix
import random

In [55]:
%matplotlib inline
from matplotlib import pylab, pyplot  # plotting

Read data.

In [56]:
data = pd.read_csv('seed_data/scores.csv')

Pivot data.

In [57]:
d = data.pivot(index='performer_id', columns='producer_id', values='score')

In [58]:
d.head()

producer_id   70       72       82       110      112      123      146      \
performer_id                                                                  
1                   0        1        0        0        0        0        0   
2                   0        1        1        1        1        1        0   
3                   0        0        1        1        1        0        0   
4                   0        1        0        1        1        1        1   
5                   0        1        0        1        0        0        0   

producer_id   158      170      184      ...  1602238  1612048  1615875  \
performer_id                             ...                              
1                   1        0        0  ...        0        0        0   
2                   0        0        1  ...        0        0        0   
3                   1        0        1  ...        0        0        0   
4                   1        1        0  ...        0        0        0   
5                   0        1        0  ...        0        0        0   

producer_id   1628850  1640674  1646412  1663511  1687278  1702507  1706060  
performer_id                                                                 
1                   0        0        0        0        0        0        0  
2                   0        0        0        0        0        0        0  
3                   0        0        0        0        0        0        0  
4                   0        0        0        0        0        0        0  
5                   0        0        0        0        0        0        0  

[5 rows x 307 columns]

This dataframe has 9,665 rows and 307 columns.  Each row corresponds to 1 performer.  Each column corresponds to a producer.

In [59]:
d.shape

(9665, 307)

In [60]:
d.tail()

producer_id   70       72       82       110      112      123      146      \
performer_id                                                                  
1734410             0        0        0        0        0        0        0   
1734423             0        0        0        0        0        0        0   
1734890             0        0        0        0        0        0        0   
1735951             0        0        0        0        0        0        0   
1736560             0        0        0        0        0        0        0   

producer_id   158      170      184      ...  1602238  1612048  1615875  \
performer_id                             ...                              
1734410             0        0        0  ...        0        0        0   
1734423             0        0        0  ...        0        0        0   
1734890             0        0        0  ...        0        0        0   
1735951             0        0        0  ...        0        0        0   
1736560             0        0        0  ...        0        0        0   

producer_id   1628850  1640674  1646412  1663511  1687278  1702507  1706060  
performer_id                                                                 
1734410             0        0        0        0        0        0        0  
1734423             0        0        0        0        0        0        0  
1734890             0        0        0        0        0        0        0  
1735951             0        0        0        0        0        0        0  
1736560             0        0        0        0        0        0        0  

[5 rows x 307 columns]

In [61]:
d.describe()

producer_id      70           72           82           110          112      \
count        9665.000000  9665.000000  9665.000000  9665.000000  9665.000000   
mean            0.030729     0.017072     0.021935     0.026280     0.022659   
std             0.172593     0.129546     0.146478     0.159976     0.148822   
min             0.000000     0.000000     0.000000     0.000000     0.000000   
25%             0.000000     0.000000     0.000000     0.000000     0.000000   
50%             0.000000     0.000000     0.000000     0.000000     0.000000   
75%             0.000000     0.000000     0.000000     0.000000     0.000000   
max             1.000000     1.000000     1.000000     1.000000     1.000000   

producer_id      123          146          158          170          184      \
count        9665.000000  9665.000000  9665.000000  9665.000000  9665.000000   
mean            0.019245     0.008898     0.016968     0.014796     0.007243   
std             0.137391     0.093914     0.129160     0.120740     0.084799   
min             0.000000     0.000000     0.000000     0.000000     0.000000   
25%             0.000000     0.000000     0.000000     0.000000     0.000000   
50%             0.000000     0.000000     0.000000     0.000000     0.000000   
75%             0.000000     0.000000     0.000000     0.000000     0.000000   
max             1.000000     1.000000     1.000000     1.000000     1.000000   

producer_id  ...      1602238      1612048      1615875      1628850  \
count        ...  9665.000000  9665.000000  9665.000000  9665.000000   
mean         ...     0.000207     0.000103     0.000103     0.000207   
std          ...     0.014384     0.010172     0.010172     0.014384   
min          ...     0.000000     0.000000     0.000000     0.000000   
25%          ...     0.000000     0.000000     0.000000     0.000000   
50%          ...     0.000000     0.000000     0.000000     0.000000   
75%          ...     0.000000     0.000000     0.000000     0.000000   
max          ...     1.000000     1.000000     1.000000     1.000000   

producer_id      1640674      1646412      1663511      1687278  1702507  \
count        9665.000000  9665.000000  9665.000000  9665.000000   9665.0   
mean            0.000310     0.000103     0.000310     0.000103      0.0   
std             0.017616     0.010172     0.017616     0.010172      0.0   
min             0.000000     0.000000     0.000000     0.000000      0.0   
25%             0.000000     0.000000     0.000000     0.000000      0.0   
50%             0.000000     0.000000     0.000000     0.000000      0.0   
75%             0.000000     0.000000     0.000000     0.000000      0.0   
max             1.000000     1.000000     1.000000     1.000000      0.0   

producer_id      1706060  
count        9665.000000  
mean            0.000103  
std             0.010172  
min             0.000000  
25%             0.000000  
50%             0.000000  
75%             0.000000  
max             1.000000  

[8 rows x 307 columns]

Show performer's 5 nearest neighbors.

In [62]:
knn = KNeighborsClassifier(n_neighbors=5)

In [63]:
list(d.index)

[1,
 2,
 3,
 4,
 5,
 13,
 18,
 20,
 21,
 22,
 32,
 34,
 38,
 42,
 44,
 45,
 46,
 56,
 58,
 59,
 63,
 64,
 67,
 68,
 69,
 70,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 85,
 86,
 88,
 89,
 92,
 93,
 100,
 103,
 105,
 107,
 108,
 110,
 111,
 112,
 115,
 117,
 118,
 119,
 122,
 123,
 124,
 126,
 127,
 130,
 132,
 135,
 140,
 141,
 142,
 144,
 146,
 147,
 148,
 149,
 150,
 156,
 157,
 158,
 160,
 161,
 162,
 167,
 169,
 170,
 173,
 176,
 177,
 178,
 179,
 180,
 182,
 183,
 184,
 186,
 190,
 191,
 192,
 195,
 196,
 197,
 198,
 201,
 202,
 203,
 204,
 206,
 207,
 209,
 210,
 211,
 215,
 216,
 217,
 219,
 220,
 221,
 223,
 224,
 225,
 228,
 229,
 230,
 232,
 233,
 236,
 238,
 240,
 244,
 245,
 249,
 250,
 251,
 255,
 257,
 260,
 262,
 266,
 267,
 270,
 272,
 276,
 278,
 282,
 286,
 287,
 288,
 289,
 291,
 293,
 296,
 301,
 303,
 304,
 305,
 306,
 307,
 310,
 311,
 312,
 315,
 316,
 318,
 319,
 320,
 322,
 323,
 326,
 327,
 328,
 329,
 330,
 332,
 334,
 335,
 337,
 338,
 339,

In [64]:
knn.fit(d.iloc[:,:], d.index)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=5, p=2,
           weights='uniform')

In [68]:
predictions = knn.predict(d.iloc[:,:])

In [69]:
joblib.dump(knn, "trained-model.pkl")

['trained-model.pkl']

List index 0 = performer, following list indices are that performer's nearest neighbors.

In [77]:
dist, ind = knn.kneighbors(d.loc[583,:].values.reshape(1, -1))
print(ind[0])

[ 300 2442  475  227  596]


In [78]:
dist

array([[0.        , 2.44948974, 2.64575131, 2.64575131, 2.64575131]])

In [79]:
[list(d.index)[i] for i in ind[0]]

[583, 25285, 1128, 444, 1461]